### ISSUE TITLE
KPH protections do not support ACG (Arbitrary Code Guard)

### ISSUE THREAD

**Issue by jxy-s** [winsiderss/systeminformer#1833]
As brought up in pull request https://github.com/winsiderss/systeminformer/pull/1832 the current protections which manipulate the CFG (Control Flow Guard) bitmask to ensure a valid call target for APC delivery are incompatible when ACG (Arbitrary Code Guard) is enabled. Enabling ACG prevents manipulation of CFG bitmask (unless certain conditions are present). We will need to consider an appropriate solution to work around this limitation with ACG enabled.

At this time, there are other issues with enabling ACG for SI. We do place some detours (hooks) on certain functions - this also conflicts with W^X from ACG and the hooks fail to set. So, ACG isn't a well supported mitigation in SI. But, we should still address the problem in the KPH mitigations. As mentioned in the PR, this prevents the client from connecting to the driver when it should be able to.

**Comment by jxy-s** [comment]
The incompatibility is originating from `Process->MitigationFlagsValues.DisableDynamicCode && !CurrentThread->DisableDynamicCodeOptOut && CFG_CALL_TARGET_VALID`. The "Disable" negating prefix here is unfortunate, but it's "technically" reasonable given the feature is intended for "disabling dynamic code".

Given that we will only really be dealing with exported call targets in for the `ApcNoopRoutine` we actually don't have to set `CFG_CALL_TARGET_VALID` as exports will already be CFG valid, but possibly Export Suppressed (ES). We can assume that it's a valid (or valid and suppressed) call target. Ultimately it was errant for me to include `CFG_CALL_TARGET_VALID` at all.

First, `ZwSetInformationVirtualMemory` checks `Process->MitigationFlagsValues.DisableDynamicCode && !CurrentThread->DisableDynamicCodeOptOut` and if so it passes `0x1` to the flags parameter of `MiCfgMarkValidEntries`. This indicates that the target process has ACG enabled and the current thread is not opting out of ACG. Interestingly, the way this is implemented in Windows, it allows you to manipulate CFG of a remote process if you yourself are ACG and have your thread opted out of ACG (even if the target remote process has ACG with thread opt-out disabled). I was goint to go this route with a fix, however (using available APIs) it requires the current process and thread have `CurrentProcess->MitigationFlagsValues.DisableDynamicCodeAllowOptOut` and `CurrentThread->DisableDynamicCodeOptOut`. This isn't realistic in all circumstances. That leaves us with DKOM to temporarily manipulate the `EPROCESS` and `ETHREAD` to permit us to do what we want. However, given `CFG_CALL_TARGET_VALID` is ultimately unnecessary, we don't have to go this route at all.

We should be fine simply removing `CFG_CALL_TARGET_VALID`. However, there is a drawback (and current bug) for CFG with ES enabled. If the call target is already a granted suppressed call target, the call will fail with `STATUS_ACCESS_DENIED`. This was already pointed out by @kkent030315 in the PR, although I think their understanding was a bit incomplete. At the end, we both arrived to a similar conclusion.

Here is the reverse engineered code of `MiCfgMarkValidEntries`, for reference:

```c
NTSTATUS __fastcall MiCfgMarkValidEntries(
        _EPROCESS *Process,
        PVOID VirtualAddress,
        SIZE_T NumberOfBytes,
        PCFG_CALL_TARGET_INFO CfgCallTargetInfo,
        ULONG NumberOfEntries,
        PULONG ReturnLength,
        ULONG Flags,                                                            // bit 1 set when Process->MitigationFlagsValues.DisableDynamicCode && !CurrentThread->DisableDynamicCodeOptOut
        _SECTION *SectionObject,
        ULONGLONG FileOffset,
        BOOLEAN AssumeValidXfgInput)                                            // effectively "IsKernelCaller"
{
//...
            cfgFlags = cfgCallTargetInfo->Flags;
            if ( (cfgFlags & 0xFFFFFFFFFFFFFFE0ui64) != 0 )                     // check for valid flags
            {
                break;
            }
                                                                                // checking for valid CFG...
            if ( (Flags & 1) != 0 && (cfgFlags & 1) != 0 )                      // Process->MitigationFlagsValues.DisableDynamicCode &&
                                                                                // !CurrentThread->DisableDynamicCodeOptOut &&
                                                                                // CFG_CALL_TARGET_VALID 
            {
                return STATUS_ACCESS_DENIED;                                    // <<< incompatbility here
            }
                                                                                // checking for valid CFG ES...
            if ( (cfgFlags & 4) != 0 )                                          // CFG_CALL_TARGET_CONVERT_EXPORT_SUPPRESSED_TO_VALID
            {
                if ( !MiIsProcessCfgExportSuppressionEnabled(Process) )         // Process->MitigationFlags2Values.ControlFlowGuardExportSuppressionEnabled
                {
                    return STATUS_ACCESS_DENIED;
                }

                currentProcess = KeGetCurrentThread()->ApcState.Process;
                if ( (CfgAddressToBitState(VirtualAddress + offset, *cfgBitMap) != 2  // is already granted suppressed call
                   || !MiIsProcessCfgExportSuppressionEnabled(currentProcess))        // currentProcess->MitigationFlags2Values.ControlFlowGuardExportSuppressionEnabled
                  && !MiValidateUserCallTarget(
                          VirtualAddress + cfgCallTargetInfo->Offset,
                          cfgBitMap) )
                {
                    return STATUS_ACCESS_DENIED;
                }
            }
                                                                                // checking for valid XFG...
            if ( (cfgFlags & 0x10) != 0                                         // CFG_CALL_TARGET_CONVERT_XFG_TO_CFG
              && (!AssumeValidXfgInput
               && (!*&Process->Vm.Instance.VmWorkingSetList[18].LastAccessClearingRemainder
                || (Process->MitigationFlags2 & 0x4000000) == 0)                // Process->MitigationFlags2Values.AuditXtendedControlFlowGuard
               || !MiIsProcessXfgEnabled(Process)                               // Process->MitigationFlags2Values.XtendedControlFlowGuard
               || !MiValidateXfgUserCallTarget(
                       VirtualAddress + cfgCallTargetInfo->Offset,
                       cfgBitMap)
               && !MiValidateUserCallTarget(
                       VirtualAddress + cfgCallTargetInfo->Offset,
                       cfgBitMap))
              || (cfgFlags & 8) != 0                                            // CFG_CALL_TARGET_VALID_XFG
              && (!AssumeValidXfgInput
               || !MiIsProcessXfgEnabled(Process)                               // Process->MitigationFlags2Values.XtendedControlFlowGuard
               || ((VirtualAddress + LOBYTE(cfgCallTargetInfo->Offset)) & 0xF) != 0
               || ((VirtualAddress + cfgCallTargetInfo->Offset) & 0xFFF) == 0) )
            {
                return STATUS_ACCESS_DENIED;
            }
// ...
```

The fist step to fix this is obviously not doing `CFG_CALL_TARGET_VALID`. That avoids the actual case of the incompatibility with ACG. However, we are then faced with the problem that if a suppressed call target is already granted access, we recieve `STATUS_ACCESS_DENIED`. @kkent030315 spotted this and handled `STATUS_ACCESS_DENIED` in the original PR, but I'm weary to just handle the `STATUS_ACCESS_DENIED` - if the return code changes in the future that assumption is likely to be broken. It would have been nice for Microsoft to specify a different error for the case where the suppressed call is already granted... The original PR also kept the `CFG_CALL_TARGET_VALID`, this would have still case a compatibility issue by not support CFG ES (the kernel would return `STATUS_ACCESS_DENIED` before granting the suppressed call access) - this would have likely resulted in an invalid call target and fast fail of the process.

As far as I'm aware, there is no a great way to query or identify that a if a call target is already a granted suppressed target. All that tracking is internal to the kernel and ntdll.

I also look at looked at the XFG path (in an effort to support XFG). Interestingly, the XFG path will effectively skip all the validation and trust the kernel caller. It should be possible to "disable XFG" (actually convert it from XFG to CFG - CFG_CALL_TARGET_CONVERT_XFG_TO_CFG) prior to granting the suppressed call target.

There is also an alternative to all this. We could choose a different ntdll routine that is imported by KernelBase. That would help ensure that the call target is valid, but it isn't technically a guarantee.

Given the facilities available to us and trying to remain conservative (not getting into DKOM, etc..). I don't think it's likely for us to get a "guarantee" here. So, I think my fix is going to perform "best effort" when setting the CFG options. And I will likely change the APC target routine to one that is imported by KernelBase. This does impose risk that if this path fails, and the call target ends up invalid, the process is going to fast fail.

**Comment by jxy-s** [comment]
I have a correction to make, `STATUS_ACCESS_DENIED` is actually only returned when the call target is invalid inside the block labeled "checking for valid CFG ES...". Here is the corrected block. I had local var mappings in IDA that were now displaying the pseudo-code code correctly.

```c
                currentProcess = KeGetCurrentThread()->ApcState.Process;
                if ( CfgAddressToBitState(VirtualAddress + offset, *cfgBitMap) == 2    // is already granted suppressed call
                  && MiIsProcessCfgExportSuppressionEnabled(currentProcess) )          // currentProcess->MitigationFlags2Values.ControlFlowGuardExportSuppressionEnabled
                {
                    cfgBitMap_1 = cfgBitMap;
                }
                else
                {
                    cfgBitMap_1 = cfgBitMap;
                    if ( !MiValidateUserCallTarget(
                              VirtualAddress + cfgCallTargetInfo->Offset,
                              cfgBitMap) )
                    {
                        return STATUS_ACCESS_DENIED;
                    }
                }
```

I verified through testing some code. So, it looks like we can rely on the return of `STATUS_ACCESS_DENIED`, at least for the case of EFG ES. The path for a kernel caller for XFG seems to still ignore the checks. Regardless, it should be reasonable to respect an errant return status in this path and just remove the `CFG_CALL_TARGET_VALID`.

**Comment by jxy-s** [comment]
Okay, the following commit will resolve the ACG compatibility problem, fix CFG ES support, and implements support for future XFG (not yet fully tested): https://github.com/winsiderss/systeminformer/commit/a3c639140649dec395d150a1bb54205f7bb93970

I discovered that the LDR is actually a bit smarter than I first expected. It didn't really matter if the call target was just "imported" (as I previously stated above). So changing the APC target made no difference in certain cases. My speculation here is that mapping and loading of the DLL is smart enough to understand that, while that export is available and possibly imported by something else, the application won't end up needing it. I reversed a little bit of the handler for invalid call targets and there are parts I haven't yet fully gawked at - but it has some extra magic to only grant where it's necessary. Research for later.

Ultimately the fix was more or less just removing `CFG_CALL_TARGET_VALID` and only granting suppressed call target (`CFG_CALL_TARGET_CONVERT_EXPORT_SUPPRESSED_TO_VALID`) where appropriate.

Will be in next driver build, until that time the workaround is to not use ACG if you want the driver functionality👍

**Comment by kkent030315** [comment]
@jxy-s Hey, thank you for the heads up on this. I appreciate the fix and the detailed explanation. ❤️ 

Essenaitlly, this issue suddenly comes up to me while I'm toggling on/off bunch of Windows features and took me a hours to debug myself. I've never touched the "Explot Protection" myself but there was a setting for SI. I suspect some Windows system automatically added it (not 100% sure though, I don't have any 3rd-party AVs & WD is permanently removed on my system either).

As you mentioned, easiest solution is to not use ACG for SI.

Anyways, I'll check out next release as possible as it released.

**Comment by jxy-s** [comment]
Resolved with latest build (`3.0.7407`) 👍 

If you have theme support enabled you will encounter this error:

![image](https://github.com/winsiderss/systeminformer/assets/11687482/87902f79-b600-43df-a40a-c6cffdf73a63)

This is expected for now as ACG will prevent us from placing hooks into our self (required for some theme support). But by in large the program should function normally with ACG enabled.
